homeassistant/components/onvif/__init__.py

```diff
@@ -1 +1,145 @@
+"""The ONVIF integration."""
+import asyncio
+
+import voluptuous as vol
+
+from homeassistant.components.ffmpeg import CONF_EXTRA_ARGUMENTS
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
+from homeassistant.const import ATTR_ENTITY_ID, CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_per_platform
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.service import async_extract_entity_ids
+
+from .const import (
+    ABSOLUTE_MOVE,
+    ATTR_CONTINUOUS_DURATION,
+    ATTR_DISTANCE,
+    ATTR_MOVE_MODE,
+    ATTR_PAN,
+    ATTR_SPEED,
+    ATTR_TILT,
+    ATTR_ZOOM,
+    CONF_PROFILE,
+    CONF_RTSP_TRANSPORT,
+    CONTINUOUS_MOVE,
+    DEFAULT_ARGUMENTS,
+    DEFAULT_PROFILE,
+    DIR_DOWN,
+    DIR_LEFT,
+    DIR_RIGHT,
+    DIR_UP,
+    DOMAIN,
+    ENTITIES,
+    RELATIVE_MOVE,
+    RTSP_TRANS_PROTOCOLS,
+    SERVICE_PTZ,
+    ZOOM_IN,
+    ZOOM_OUT,
+)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["camera"]
+
+SERVICE_PTZ_SCHEMA = vol.Schema(
+    {
+        ATTR_ENTITY_ID: cv.entity_ids,
+        vol.Optional(ATTR_PAN): vol.In([DIR_LEFT, DIR_RIGHT]),
+        vol.Optional(ATTR_TILT): vol.In([DIR_UP, DIR_DOWN]),
+        vol.Optional(ATTR_ZOOM): vol.In([ZOOM_OUT, ZOOM_IN]),
+        ATTR_MOVE_MODE: vol.In([CONTINUOUS_MOVE, RELATIVE_MOVE, ABSOLUTE_MOVE]),
+        vol.Optional(ATTR_CONTINUOUS_DURATION, default=0.5): cv.small_float,
+        vol.Optional(ATTR_DISTANCE, default=0.1): cv.small_float,
+        vol.Optional(ATTR_SPEED, default=0.5): cv.small_float,
+    }
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the ONVIF component."""
+    # Create PTZ service
+    async def async_handle_ptz(service):
+        """Handle PTZ service call."""
+        pan = service.data.get(ATTR_PAN)
+        tilt = service.data.get(ATTR_TILT)
+        zoom = service.data.get(ATTR_ZOOM)
+        distance = service.data[ATTR_DISTANCE]
+        speed = service.data[ATTR_SPEED]
+        move_mode = service.data.get(ATTR_MOVE_MODE)
+        continuous_duration = service.data[ATTR_CONTINUOUS_DURATION]
+        all_cameras = hass.data[DOMAIN][ENTITIES]
+        entity_ids = await async_extract_entity_ids(hass, service)
+        target_cameras = []
+        if not entity_ids:
+            target_cameras = all_cameras
+        else:
+            target_cameras = [
+                camera for camera in all_cameras if camera.entity_id in entity_ids
+            ]
+        for camera in target_cameras:
+            await camera.async_perform_ptz(
+                pan, tilt, zoom, distance, speed, move_mode, continuous_duration
+            )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
+    )
+
+    # Import from yaml
+    configs = {}
+    for p_type, p_config in config_per_platform(config, "camera"):
```

Comment:
Wouldn't it be easier to just do this inside the `async_setup_platform` of the Onvif camera platform?

Reply:
The uniqueness is a combination of host and profile. Previously for NVRs someone would configure multiple onvif camera platforms Where the of my difference was the profile index. The way I’m solving this moving forward is to have a single config entry per device that creates multiple camera entities based on the configured profiles.